controllers/storagecluster/provider_server_test.go

```diff
@@ -0,0 +1,169 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	"github.com/red-hat-storage/ocs-operator/controllers/util"
+)
+
+func TestOcsProviderServerEnsureCreated(t *testing.T) {
+
+	t.Run("Ensure that Deployment,Service,Secret is created when AllowRemoteStorageConsumers is enabled", func(t *testing.T) {
+
+		r, instance := createSetupForOcsProviderTest(t, true)
+
+		obj := &ocsProviderServer{}
+		_, err := obj.ensureCreated(r, instance)
+
+		assert.Error(t, err)
+		assert.Equal(t, "Deployment ocs-provider-server is not ready", err.Error())
+
+		deployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{Name: ocsProviderServerName},
+		}
+		assertFound(t, r.Client, deployment)
+		expectedDeployment := GetProviderAPIServerDeployment(instance)
+		assert.Equal(t, deployment.Spec, expectedDeployment.Spec)
+
+		service := &corev1.Service{
+			ObjectMeta: metav1.ObjectMeta{Name: ocsProviderServerName},
+		}
+		assertFound(t, r.Client, service)
+		expectedService := GetProviderAPIServerService(instance)
+		assert.Equal(t, expectedService.Spec, service.Spec)
+
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{Name: ocsProviderServerName},
+		}
+		assertFound(t, r.Client, secret)
+	})
+
+	t.Run("Ensure that Deployment,Service,Secret is not created when AllowRemoteStorageConsumers is disabled", func(t *testing.T) {
+
+		r, instance := createSetupForOcsProviderTest(t, false)
+
+		obj := &ocsProviderServer{}
+		_, err := obj.ensureCreated(r, instance)
+
+		assert.NoError(t, err)
+
+		assertNotFoundProviderResources(t, r.Client)
+	})
+}
+
+func TestOcsProviderServerEnsureDeleted(t *testing.T) {
+
+	t.Run("Ensure that Deployment,Service,Secret is deleted when AllowRemoteStorageConsumers is disabled", func(t *testing.T) {
+
+		r, instance := createSetupForOcsProviderTest(t, true)
+		obj := &ocsProviderServer{}
+		// create resources and ignore error as it should be tested via TestOcsProviderServerEnsureCreated
+		_, _ = obj.ensureCreated(r, instance)
+
+		instance.Spec.AllowRemoteStorageConsumers = false
+		// the resources will be deleted through the ensureCreated func as we are not in the deletion phase
+		_, err := obj.ensureCreated(r, instance)
+		assert.NoError(t, err)
+
+		assertNotFoundProviderResources(t, r.Client)
+	})
+
+	t.Run("Ensure that Deployment,Service,Secret is deleted while uninstalling", func(t *testing.T) {
+
+		r, instance := createSetupForOcsProviderTest(t, true)
+		obj := &ocsProviderServer{}
+		// create resources and ignore error as it should be tested via TestOcsProviderServerEnsureCreated
+		_, _ = obj.ensureCreated(r, instance)
+
+		_, err := obj.ensureDeleted(r, instance)
+		assert.NoError(t, err)
+
+		assertNotFoundProviderResources(t, r.Client)
+	})
+}
+
+func assertFound(t *testing.T, cli client.Client, object client.Object) {
+
+	key := client.ObjectKeyFromObject(object)
+
+	err := cli.Get(context.TODO(), key, object)
+	assert.NoError(t, err)
+}
```

Comment:
Why do we need this util when we can just do this (in the call site) instead?
```suggestion
assert.NoError(t, cli.Get(context.TODO(), client.ObjectKeyFromObject(object), object))
```